File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudOpusSegmentDecoder.java
package com.sedmelluq.discord.lavaplayer.source.soundcloud;

import com.sedmelluq.discord.lavaplayer.container.ogg.OggPacketInputStream;
import com.sedmelluq.discord.lavaplayer.container.ogg.OggTrackBlueprint;
import com.sedmelluq.discord.lavaplayer.container.ogg.OggTrackHandler;
import com.sedmelluq.discord.lavaplayer.container.ogg.OggTrackLoader;
import com.sedmelluq.discord.lavaplayer.tools.io.SeekableInputStream;
import com.sedmelluq.discord.lavaplayer.track.playback.AudioProcessingContext;

import java.io.IOException;
import java.util.function.Supplier;

public class SoundCloudOpusSegmentDecoder implements SoundCloudSegmentDecoder {
    private final Supplier<SeekableInputStream> nextStreamProvider;
    private OggPacketInputStream lastJoinedStream;
    private OggTrackBlueprint blueprint;

    public SoundCloudOpusSegmentDecoder(Supplier<SeekableInputStream> nextStreamProvider) {
        this.nextStreamProvider = nextStreamProvider;
    }

    @Override
    public void prepareStream(boolean beginning) throws IOException {
        OggPacketInputStream stream = obtainStream();

        if (beginning) {
            OggTrackBlueprint newBlueprint = OggTrackLoader.loadTrackBlueprint(stream);

            if (blueprint == null) {
                if (newBlueprint == null) {
                    throw new IOException("No OGG track detected in the stream.");
                }

                blueprint = newBlueprint;
            }
        } else {
            stream.startNewTrack();
        }
    }

    @Override
    public void resetStream() throws IOException {
        if (lastJoinedStream != null) {
            lastJoinedStream.close();
            lastJoinedStream = null;
        }
    }

    @Override
    public void playStream(
        AudioProcessingContext context,
        long startPosition,
        long desiredPosition
    ) throws InterruptedException, IOException {
        try (OggTrackHandler handler = blueprint.loadTrackHandler(obtainStream())) {
            handler.initialise(
                context,
                startPosition,
                desiredPosition
            );

            handler.provideFrames();
        }
    }

    @Override
    public void close() throws Exception {
        resetStream();
    }

    private OggPacketInputStream obtainStream() {
        if (lastJoinedStream == null) {
            lastJoinedStream = new OggPacketInputStream(nextStreamProvider.get(), true);
        }

        return lastJoinedStream;
    }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudOpusSegmentDecoder.kt
package com.sedmelluq.discord.lavaplayer.source.soundcloud

import com.sedmelluq.discord.lavaplayer.container.ogg.OggPacketInputStream
import com.sedmelluq.discord.lavaplayer.container.ogg.OggTrackBlueprint
import com.sedmelluq.discord.lavaplayer.container.ogg.OggTrackLoader
import com.sedmelluq.discord.lavaplayer.tools.io.SeekableInputStream
import com.sedmelluq.discord.lavaplayer.track.playback.AudioProcessingContext
import java.io.IOException
import java.util.function.Supplier

class SoundCloudOpusSegmentDecoder(private val nextStreamProvider: Supplier<SeekableInputStream>) :
    SoundCloudSegmentDecoder {
    private var lastJoinedStream: OggPacketInputStream? = null
    private var blueprint: OggTrackBlueprint? = null

    @Throws(IOException::class)
    override fun prepareStream(beginning: Boolean) {
        val stream = obtainStream()
        if (beginning) {
            val newBlueprint = OggTrackLoader.loadTrackBlueprint(stream)
            if (blueprint == null) {
                if (newBlueprint == null) {
                    throw IOException("No OGG track detected in the stream.")
                }
                blueprint = newBlueprint
            }
        } else {
            stream.startNewTrack()
        }
    }

    @Throws(IOException::class)
    override fun resetStream() {
        if (lastJoinedStream != null) {
            lastJoinedStream!!.close()
            lastJoinedStream = null
        }
    }

    @Throws(InterruptedException::class, IOException::class)
    override fun playStream(context: AudioProcessingContext, startPosition: Long, desiredPosition: Long) {
        blueprint!!.loadTrackHandler(obtainStream()).use { handler ->
            handler.initialise(
                context,
                startPosition,
                desiredPosition
            )
            handler.provideFrames()
        }
    }

    @Throws(Exception::class)
    override fun close() {
        resetStream()
    }

    private fun obtainStream(): OggPacketInputStream {
        if (lastJoinedStream == null) {
            lastJoinedStream = OggPacketInputStream(nextStreamProvider.get(), true)
        }

        return lastJoinedStream as OggPacketInputStream
    }
}
